File: FestivalJava/Networking/src/main/java/rpcprotocol/FestivalServerRpcProxy.java
package rpcprotocol;

import AppServices.IFestivalClient;
import AppServices.IFestivalServer;
import Domain.Artist;
import Domain.Show;
import Domain.User;
import Validation.Exceptions.ServiceException;
import Validation.Exceptions.ValidatorException;
import dto.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.rmi.RemoteException;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Created by Sebi on 29-Mar-17.
 */
public class FestivalServerRpcProxy implements IFestivalServer {
    //host
    private String host;

    //connection
    private Socket connection;
    private int port;
    private ObjectInputStream inputStream;
    private ObjectOutputStream outputStream;

    //reference to client
    private IFestivalClient client;

    //server responses
    private BlockingQueue<Response> queueResponses;
    //indicates conection status
    private volatile boolean finished;

    //constructor
    public FestivalServerRpcProxy(String host, int port){
        this.host = host;
        this.port = port;
        queueResponses = new LinkedBlockingQueue<>();
    }

    @Override
    public List<Artist> getArtists() throws ServiceException{
        Request request = new Request.Builder().type(RequestType.GET_ARTISTS).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR){
            String msg = (String) response.getData();
            throw new ServiceException(msg);
        }
        List<ArtistDTO> artistDTOs = (List<ArtistDTO>) response.getData();
        return DTOUtils.getListArtistFromDTO(artistDTOs);
    }

    @Override
    public List<Show> getShowsForArtist(Integer idArtist) throws ServiceException {
        Request request = new Request.Builder().type(RequestType.GET_SHOWS_FOR_ARTIST).build();
        request.setData(idArtist);
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR){
            String msg = (String) response.getData();
            throw new ServiceException(msg);
        }
        List<ShowDTO> showDTOs = (List<ShowDTO>) response.getData();
        return DTOUtils.getListShowFromDTO(showDTOs);
    }

    @Override
    public List<Show> getShowsForDate(String date) throws ServiceException{
        Request request = new Request.Builder().type(RequestType.GET_SHOWS_FOR_DATE).build();
        request.setData(date);
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR){
            String msg = (String) response.getData();
            throw new ServiceException(msg);
        }
        List<ShowDTO> showDTOs = (List<ShowDTO>) response.getData();
        return DTOUtils.getListShowFromDTO(showDTOs);
    }

    @Override
    public void buyTicketsForShow(Integer idShow, String clientName, Integer numberOfTickets, String username) throws ServiceException, ValidatorException {

        BuyTicketsDTO buyTicketsDTO = new BuyTicketsDTO(idShow, clientName, numberOfTickets, username);
        Request request = new Request.Builder().type(RequestType.BUY_TICKETS).data(buyTicketsDTO).build();

        sendRequest(request);
        Response response = readResponse();

        if (response.getType() == ResponseType.ERROR){
            String msg = (String) response.getData();
            throw new ServiceException(msg);
        }
    }

    @Override
    public void login(User user, IFestivalClient client) throws ServiceException {
        initializeConnection();
        UserDTO userDTO = DTOUtils.getUserDTO(user);
        Request request = new Request.Builder().type(RequestType.LOGIN).data(userDTO).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.OK){
            this.client = client;
            return;
        }
        if (response.getType() == ResponseType.ERROR){
            String message = (String)response.getData();
            throw new ServiceException(message);
        }
    }

    @Override
    public void logout(User user, IFestivalClient client) throws ServiceException {
        UserDTO userDTO = DTOUtils.getUserDTO(user);
        Request request = new Request.Builder().type(RequestType.LOGOUT).data(userDTO).build();
        sendRequest(request);
        Response response = readResponse();
        closeConnection();
        if (response.getType() == ResponseType.ERROR){
            throw new ServiceException("Error logout");
        }
    }

    /*
    Sends a request to server
     */
    private void sendRequest(Request request) throws ServiceException{
        try{
            System.out.println("Sending request " + request.getType());
            outputStream.writeObject(request);
            outputStream.flush();
        } catch (IOException e) {
            throw new ServiceException("Error sending request");
        }
    }

    /*
    Reads the first response from the queue
     */
    private Response readResponse(){
        Response response = null;
        try{
            response = queueResponses.take();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return response;
    }

    /*
    Initialize connection to server
     */
    private void initializeConnection() throws ServiceException {
        try{
            connection = new Socket(host, port);
            outputStream = new ObjectOutputStream(connection.getOutputStream());
            outputStream.flush();
            inputStream = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
        } catch (IOException e) {
            throw new ServiceException("Connection to server refused");
        }
    }

    /*
    Closes the connection
     */
    private void closeConnection(){
        finished = true;
        try{
            inputStream.close();
            outputStream.close();
            connection.close();
            client = null;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    //starts the reader thread
    private void startReader(){
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }

    //checks if a response is update type
    private boolean isUpdate(Response response){
        return response.getType().equals(ResponseType.SHOWS_UPDATED);
    }

    //handles updates from server
    private void handleUpdate(Response response){
        //show values updated after transaction
        if (response.getType() == ResponseType.SHOWS_UPDATED){
            try {
                ShowDTO showDTO = (ShowDTO)response.getData();
                Show show = DTOUtils.getShowFromDTO(showDTO);
                client.showUpdated(show);
            } catch (ServiceException | RemoteException e) {
                e.printStackTrace();
            }
        }
    }

    /*
    Reader thread
    reads responses from server
     */
    private class ReaderThread implements Runnable{

        @Override
        public void run() {
            while (!finished){
                try{
                    System.out.println("Tries to read..");
                    Object response = inputStream.readObject();
                    System.out.println("Response received " + ((Response)response).getType());
                    if (isUpdate((Response) response)){
                        handleUpdate((Response) response);
                    } else {
                        try{
                            queueResponses.put((Response) response);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (IOException | ClassNotFoundException ex){
                    System.out.println("Client reader stopped");
                    finished = true;
                }
            }
        }
    }
}

File: FestivalC#/Festival/Networking/rpcprotocol/FestivalServerRpcProxy.cs
using Festival.Service;
using Festival.Service.AppServices;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Festival.Model;
using Networking.dto;
using Services.Validation.Exceptions;
using System.Collections;

namespace Networking.rpcprotocol
{
    public class FestivalServerRpcProxy : IFestivalServer
    {
        private string host;
        private int port;
        private NetworkStream stream;
        private IFormatter formatter;
        private TcpClient connection;

        private Queue<Response> responses;
        private volatile bool finished;
        private EventWaitHandle waitHandle;

        private IFestivalClient client;

        public FestivalServerRpcProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<Response>();
        }

        //sends a request to server
        private void sendRequest(Request request)
        {
            try
            {
                formatter.Serialize(stream, request);
                stream.Flush();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        //reads a response from queue
        private Response readResponse()
        {
            Response response = null;
            try
            {
                waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();
                }
            } catch(Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            return response;
        }

        //initialize the connection
        private void initializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                waitHandle = new AutoResetEvent(false);
                startReader();
            } catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        //close the connection
        private void closeConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                waitHandle.Close();
                client = null; 
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        //starts the reader thread
        private void startReader()
        {
            Thread tw = new Thread(run);
            tw.Start();
        }

        //checks if response type is update
        private bool isUpdate(Response response)
        {
            if (response.Type == ResponseType.SHOWS_UPDATED)
            {
                return true;
            }
            return false;
        }

        //background reader thread run method
        public virtual void run()
        {
            while (!finished)
            {
                try
                {
                    Response response = (Response)formatter.Deserialize(stream);
              
                    Console.WriteLine("Response received " + response);
                    if (isUpdate(response))
                    {

                    } else
                    {
                        lock (responses)
                        {
                            responses.Enqueue(response);
                        }
                        waitHandle.Set();
                    }
                } catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                }
            }
        }



        public List<Artist> getArtists()
        {
            List<Artist> artists = new List<Artist>();
            sendRequest(new Request(RequestType.GET_ARTISTS, null));
            Response response = readResponse();
            if (response.Type == ResponseType.ERROR)
            {
                throw new ServiceException((string)response.Data);
            }
            List<ArtistDTO> artistDTO = ((IEnumerable)response.Data).Cast<ArtistDTO>().ToList();
            return DTOUtils.getListArtistFromDTO(artistDTO);
        }

        public List<Show> getShowsForArtist(int idArtist)
        {
            List<Show> shows = new List<Show>();
            sendRequest(new Request(RequestType.GET_SHOWS_FOR_ARTIST, idArtist));
            Response response = readResponse();
            if (response.Type == ResponseType.ERROR)
            {
                throw new ServiceException((string)response.Data);
            }
            List<ShowDTO> showDTO = ((IEnumerable)response.Data).Cast<ShowDTO>().ToList();
            return DTOUtils.getListShowFromDTO(showDTO);
        }

        public List<Show> getShowsForDate(string date)
        {
            List<Show> shows = new List<Show>();
            sendRequest(new Request(RequestType.GET_SHOWS_FOR_DATE, date));
            Response response = readResponse();
            if (response.Type == ResponseType.ERROR)
            {
                throw new ServiceException((string)response.Data);
            }
            List<ShowDTO> showDTO = ((IEnumerable)response.Data).Cast<ShowDTO>().ToList();
            return DTOUtils.getListShowFromDTO(showDTO);
        }

        public void buyTicketsForShow(int idShow, string clientName, int numberOfTickets, string username)
        {
            throw new NotImplementedException();
        }

        //login request
        public void login(User user, IFestivalClient client)
        {
            initializeConnection();
            UserDTO userDTO = DTOUtils.getUserDTO(user);
            sendRequest(new Request(RequestType.LOGIN, userDTO));
            Response response = readResponse();
            if (response.Type == ResponseType.OK)
            {
                this.client = client;
                return;
            }
            if (response.Type == ResponseType.ERROR)
            {
                throw new ServiceException((string)response.Data);
            }
        }

        //logout request
        public void logout(User user, IFestivalClient client)
        {
            UserDTO userDTO = DTOUtils.getUserDTO(user);
            sendRequest(new Request(RequestType.LOGOUT, userDTO));
            Response response = readResponse();

            if (response.Type == ResponseType.OK)
            {
                this.client = null;
                closeConnection();
                return;
            }
            if (response.Type == ResponseType.ERROR)
            {
                throw new ServiceException((string)response.Data);
            }
        }
    }
}
